keps/sig-network/2136-np-version-status/README.md

```diff
@@ -0,0 +1,858 @@
+were actually parsed.
+
+[incorrect CIDR notation]: https://github.com/kubernetes/kubernetes/pull/94484
+
+#### Story 5 - Changing an Alpha NetworkPolicy API
+
+As a kubernetes developer, I want to make a change to a previously-added
+alpha NetworkPolicy feature, in a way that does not cause problems for
+users in clusters with varying apiserver and network plugin versions.
+
+### Notes/Constraints/Caveats
+
+#### Feature Gates and NetworkPolicy
+
+Alpha APIs create an additional problem for NetworkPolicy, due to the
+three-way (user/apiserver/plugin) versioning split; if an alpha field
+is added in 1.21 but then changed in 1.22, then a 1.22 cluster with a
+network plugin implementing the 1.21 version would probably be unable
+to use either version of the feature. We might want to establish a
+somewhat more restrictive set of alpha API rules for NetworkPolicy
+changes than are used for other API types.
+
+#### Determining Whether a Network Plugin Implements NetworkPolicyStatus
+
+If a user is waiting to see the status of a newly-created NetworkPolicy,
+there is no entirely-reliable way to distinguish "the plugin has not yet
+set `status` but will soon" from "the plugin doesn't know about `status`
+and is never going to set it".
+
+It's not clear how big a problem this is, especially if we suggest that
+implementations should create an "empty" `status` right away if it's
+going to take them a while to determine the final `status`.
+
+#### Distributed and Delegating NetworkPolicy Implementations
+
+Some NetworkPolicy implementations do not operate from a single
+controller, but instead operate independently on every node (much like
+how kube-proxy implements Service rules). In this case it is not clear
+who would be responsible for updating `NetworkPolicyStatus`, and it may
+not be easy for any part of the system to know with certainty that a
+given NetworkPolicy was fully in effect across the cluster.
+
+In other cases, a plugin may implement NetworkPolicy in terms of some
+other lower-level policy engine (implemented by a cloud, perhaps), and
+may work by simply translating NetworkPolicy rules into rules for this
+other engine. As in the distributed case, it may not be possible for
+such a NetworkPolicy implementation to know for sure when a policy was
+fully in effect.
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+Add a new field to `networkingv1.NetworkPolicySpec`:
+
+```
+type NetworkPolicySpec struct {
+	...
+
+	// minVersion indicates the Kubernetes release version corresponding to
+	// the NetworkPolicy functionality required by this policy. If this is
+	// explicitly specified by the user, the apiserver will require that the policy
+	// conforms to the specified version. If it is not specified, the apiserver
+	// will fill in the correct minVersion based on the features used by the policy.
+	// +optional
+	MinVersion NetworkPolicyVersion `json:"minVersion,omitempty" protobuf:"bytes,5,name=minVersion"`
```

Comment:
Absent a general mechanism to do this automatically, I don't think this is really going to work.

Reply:
We have to modify `ValidateNetworkPolicySpec` every time we add a new feature anyway. It would not be difficult at all to modify it to keep track of what features it has seen at the same time.

eg, if there is a `NetworkPolicyPort.Protocol`, it must be one of the valid `v1.Protocol` values. _And_, if that value is `"SCTP"` then the `minVersion` is now at least `"1.12"`

if there is a `NetworkPolicyPort.EndPort`, then there must also be `NetworkPolicyPort.Port`, and `Port` must be smaller than `EndPort`. And `minVersion` is now at least `"1.21"`

etc

Reply:
Anything is possible, it's just code.  But the code around that becomes complicated quickly.  "I saw field foo, so it must be at least 1.22, but it had value bar which means it is at least 1.23.

I'm not even against the idea overall, it might be cleaner to validate in discrete steps.  I just a) don't want to open code it; and b) don't want to walk this road alone.  This is what IDLs are for.